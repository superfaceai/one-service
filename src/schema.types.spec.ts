import {
  NormalizedProfileSettings,
  SecurityType,
  HttpScheme,
  ProviderJson,
  ApiKeyPlacement,
} from '@superfaceai/ast';
import {
  ObjectStructure,
  PrimitiveStructure,
  StructureType,
} from '@superfaceai/parser';
import {
  GraphQLBoolean,
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLString,
} from 'graphql';
import {
  enumType,
  generateProfileConfig,
  generateProfileTypes,
  generateRootType,
  generateStructureResultType,
  generateUseCaseFieldConfig,
  generateProfileProviderOptionInputType,
  generateUseCaseProviderParametersFields,
  generateUseCaseSecurityFields,
  outputType,
  primitiveType,
  ProvidersJsonRecord,
  scalarType,
  prepareProviderConfigTypeMap,
} from './schema.types';
import {
  expectSchema,
  getProfileOutput,
  parseProfileFixture,
} from './spec_helpers';

const objectStructure: ObjectStructure = {
  kind: 'ObjectStructure',
  title: 'Object title',
  description: 'Object description',
  fields: {
    primitive_string: {
      kind: 'PrimitiveStructure',
      type: 'string',
      title: 'Primitive',
      description: 'String',
    },
    primitive_number: {
      kind: 'PrimitiveStructure',
      type: 'number',
      title: 'Primitive',
      description: 'Number',
    },
    primitive_boolean: {
      kind: 'PrimitiveStructure',
      type: 'boolean',
      title: 'Primitive',
      description: 'boolean',
    },
    nested_object: {
      kind: 'ObjectStructure',
      title: 'Object',
      description: 'Nested',
      fields: {
        field: {
          kind: 'PrimitiveStructure',
          type: 'string',
          title: 'Nested field title',
          description: 'Nested field description',
        },
      },
    },
    list: {
      kind: 'ListStructure',
      value: {
        kind: 'NonNullStructure',
        value: {
          kind: 'PrimitiveStructure',
          type: 'string',
          title: 'List item title',
          description: 'List item description',
        },
      },
    },
    enum: {
      kind: 'EnumStructure',
      enums: [{ name: 'ONE', value: 'one' }, { value: 'TWO' }],
      title: 'Enum title',
      description: 'Enum description',
    },
  },
};

describe('schema.types', () => {
  const profileSettings: NormalizedProfileSettings = {
    version: '1.0.0',
    defaults: {},
    priority: ['mock'],
    providers: {
      mock: {
        defaults: {},
      },
      superface: {
        defaults: {},
      },
      'gql_bad-name': {
        defaults: {},
      },
    },
  };

  const providersJsons: ProvidersJsonRecord = {
    mock: {
      name: 'mock',
      defaultService: 'default',
      services: [
        {
          id: 'default',
          baseUrl: 'loop.back',
        },
      ],
    },
    superface: {
      name: 'superface',
      defaultService: 'default',
      services: [
        {
          id: 'default',
          baseUrl: 'https://superface.test',
        },
      ],
      securitySchemes: [
        {
          id: 'api_key',
          type: SecurityType.APIKEY,
          in: ApiKeyPlacement.HEADER,
          name: 'X-API-KEY',
        },
        {
          id: 'basic',
          type: SecurityType.HTTP,
          scheme: HttpScheme.BASIC,
        },
        {
          id: 'bearer_token',
          type: SecurityType.HTTP,
          scheme: HttpScheme.BEARER,
          bearerFormat: 'JWT',
        },
        {
          id: 'digest',
          type: SecurityType.HTTP,
          scheme: HttpScheme.DIGEST,
          statusCode: 401,
          authorizationHeader: 'Authorization',
          challengeHeader: 'www-authenticate',
        },
      ],
      parameters: [
        {
          name: 'accessToken',
        },
      ],
    },
    'gql_bad-name': {
      name: 'gql_bad-name',
      defaultService: 'default',
      services: [
        {
          id: 'default',
          baseUrl: 'https://superface.test',
        },
      ],
    },
  };

  const providerConfigTypeMap = prepareProviderConfigTypeMap(providersJsons);

  describe('generateProfileTypes', () => {
    it('skips QueryType if no safe usecase is present', async () => {
      const profileAst = await parseProfileFixture('unsafe_only');

      const result = await generateProfileTypes(
        'ScopeName',
        profileAst,
        profileSettings,
        providerConfigTypeMap,
      );

      expect(result.QueryType).toBeUndefined();
    });

    it('skips MutationType if only safe usecases are present', async () => {
      const profileAst = await parseProfileFixture('safe_only');

      const result = await generateProfileTypes(
        'ScopeName',
        profileAst,
        profileSettings,
        providerConfigTypeMap,
      );

      expect(result.MutationType).toBeUndefined();
    });
  });

  describe('generateUseCaseFieldConfig', () => {
    const ProfileProviderOptionType = new GraphQLInputObjectType({
      name: 'ScopeNameProviderOption',
      fields: {
        provider: {
          type: new GraphQLInputObjectType({
            name: 'ScopeNameProviderConfig',
            fields: {
              active: { type: GraphQLBoolean },
            },
          }),
        },
      },
    });

    describe('valid profile', () => {
      it('creates field config with arguments, resolver and description', async () => {
        const profileAst = await parseProfileFixture('profile');
        const profileOutput = await getProfileOutput('profile', profileAst);

        const config = generateUseCaseFieldConfig(
          'ScopeName',
          profileAst,
          profileOutput.usecases[0],
          ProfileProviderOptionType,
        );

        expect(config).toMatchSnapshot();
      });
    });

    describe('usecase with empty input', () => {
      it('creates arguments without input', async () => {
        const profileAst = await parseProfileFixture(
          'profile_with_empty_input_structure',
        );
        const profileOutput = await getProfileOutput(
          'profile_with_empty_input_structure',
          profileAst,
        );
        const config = generateUseCaseFieldConfig(
          'ScopeName',
          profileAst,
          profileOutput.usecases[0],
          ProfileProviderOptionType,
        );

        expect(config).toMatchSnapshot();
      });
    });
  });

  describe('generateProfileConfig', () => {
    it('creates config with description and resolve function', async () => {
      const profileAst = await parseProfileFixture('profile');

      expect(
        generateProfileConfig(
          new GraphQLObjectType({ name: 'Test', fields: {} }),
          profileAst,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('generateStructureResultType', () => {
    it('creates ScopeNameResult with description and result field', async () => {
      const profileOutput = await getProfileOutput('profile');
      expectSchema(
        generateStructureResultType(
          'ScopeNameResult',
          profileOutput.usecases[0].result as StructureType,
        ),
      );
    });

    it('creates ScopeNameResult as None value', async () => {
      const profileOutput = await getProfileOutput('no_result');
      expectSchema(
        generateStructureResultType(
          'ScopeNameResult',
          profileOutput.usecases[0].result as StructureType,
        ),
      );
    });

    it('creates ScopeNameResult as None value if result is empty object', async () => {
      const profileOutput = await getProfileOutput('profile_with_empty_result');
      expectSchema(
        generateStructureResultType(
          'ScopeNameResult',
          profileOutput.usecases[0].result as StructureType,
        ),
      );
    });

    it('creates ScopeNameResult if result contains nested empty object', async () => {
      const profileOutput = await getProfileOutput(
        'profile_with_empty_nested_object_in_result',
      );
      expectSchema(
        generateStructureResultType(
          'ScopeNameResult',
          profileOutput.usecases[0].result as StructureType,
        ),
      );
    });
  });

  describe('prepareProviderConfigTypeMap', () => {
    it('creates map of provider config types', () => {
      expect(prepareProviderConfigTypeMap(providersJsons)).toMatchSnapshot();
    });
  });

  describe('generateProfileProviderOptionInputType', () => {
    it('creates provider object with security and parameters field', () => {
      expectSchema(
        generateProfileProviderOptionInputType(
          'Test',
          profileSettings,
          providerConfigTypeMap,
        ),
      );
    });
  });

  describe('generateUseCaseProviderParametersFields', () => {
    it('generates provider parameters for configured providers', () => {
      expect(
        generateUseCaseProviderParametersFields(providersJsons, [
          'superface',
          'mock',
        ]),
      ).toMatchSnapshot();
    });

    it('returns undefined when no parameters are configured', () => {
      const emptyProvider: Omit<ProviderJson, 'name'> = {
        defaultService: 'default',
        services: [],
        securitySchemes: [],
        parameters: [],
      };
      expect(
        generateUseCaseProviderParametersFields(
          {
            foo: { name: 'foo', ...emptyProvider },
            bar: { name: 'bar', ...emptyProvider },
          },
          ['foo', 'bar'],
        ),
      ).toBeUndefined();
    });
  });

  describe('generateUseCaseSecurityFields', () => {
    it('generates security fields for configured providers', () => {
      expect(
        generateUseCaseSecurityFields(providersJsons, ['mock', 'superface']),
      ).toMatchSnapshot();
    });
  });

  describe('generateRootType', () => {
    it('create Query type with profiles', () => {
      expectSchema(
        generateRootType('Query', [
          new GraphQLObjectType({
            name: 'ProfileOne',
            fields: {
              field: {
                type: GraphQLString, // Shortcut for tests, structure here is more complicated
              },
            },
          }),
          new GraphQLObjectType({
            name: 'Scope',
            fields: {
              ProfileTwo: {
                type: new GraphQLObjectType({
                  name: 'ScopeProfileTwo',
                  fields: {
                    UseCase: {
                      type: GraphQLString,
                    },
                  },
                }),
              },
            },
          }),
        ]),
      );
    });
  });

  describe('primitveType', () => {
    it('throws error if type is unknown', () => {
      expect(() =>
        primitiveType('', { kind: 'PrimitiveStructure' } as PrimitiveStructure),
      ).toThrowError();
    });
  });

  describe('scalarType', () => {
    it('returns GraphQLScalarType', () => {
      expectSchema(
        scalarType('TestScalar', {
          kind: 'ScalarStructure',
          title: 'Scalar title',
          description: 'Scalar description',
        }),
      );
    });
  });

  describe('enumType', () => {
    it('returns GraphQLEnumType', () => {
      expectSchema(
        enumType('TestEnum', {
          kind: 'EnumStructure',
          enums: [{ name: 'O_N_E', value: 'ONE' }, { value: 'TWO' }],
          title: 'Enum title',
          description: 'Enum description',
        }),
      );
    });
  });

  describe('outputType', () => {
    it('returns GraphQLObjectType', () => {
      expectSchema(outputType('MyObjectType', objectStructure));
    });
  });

  describe('inputType', () => {
    it('returns GraphQLInputType', () => {
      expectSchema(outputType('MyObjectType', objectStructure));
    });
  });
});
