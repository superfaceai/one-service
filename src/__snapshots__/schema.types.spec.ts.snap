// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema.types enumType returns GraphQLEnumType 1`] = `
""""
Enum title
Enum description
"""
enum TestEnum {
  O_N_E
  TWO
}"
`;

exports[`schema.types generateProfileConfig creates config with description and resolve function 1`] = `
{
  "description": "Full profile
Full profile to test all generation functionality",
  "resolve": [Function],
  "type": "Test",
}
`;

exports[`schema.types generateProfileProviderOptionInputType creates provider object with security and parameters field 1`] = `
""""Provider configuration for OneSDK perform"""
input Test {
  """Provider mock configuration"""
  mock: TestMockConfig

  """Provider superface configuration"""
  superface: TestSuperfaceConfig
}

input TestMockConfig {
  active: Boolean
}

input TestSuperfaceConfig {
  active: Boolean
  parameters: TestSuperfaceProviderParameters
  security: TestSuperfaceProviderSecurity
}

"""Provider-specific parameters"""
input TestSuperfaceProviderParameters {
  """Parameter accepted by superface"""
  accessToken: String
}

"""Provider-specific security"""
input TestSuperfaceProviderSecurity {
  """Security accepted by superface"""
  api_key: SuperfaceApiKeySecurityValues

  """Security accepted by superface"""
  basic: SuperfaceBasicSecurityValues

  """Security accepted by superface"""
  bearer_token: SuperfaceBearerTokenSecurityValues

  """Security accepted by superface"""
  digest: SuperfaceDigestSecurityValues
}

input SuperfaceApiKeySecurityValues {
  apikey: String
}

input SuperfaceBasicSecurityValues {
  username: String
  password: String
}

input SuperfaceBearerTokenSecurityValues {
  token: String
}

input SuperfaceDigestSecurityValues {
  username: String
  password: String
}"
`;

exports[`schema.types generateRootType create Query type with profiles 1`] = `
"type Query {
  ProfileOne: ProfileOne
  Scope: Scope
}

type ProfileOne {
  field: String
}

type Scope {
  ProfileTwo: ScopeProfileTwo
}

type ScopeProfileTwo {
  UseCase: String
}"
`;

exports[`schema.types generateStructureResultType creates ScopeNameResult with description and result field 1`] = `
""""
Wrapping type to handle many possible types returned as result by OneSDK
"""
type ScopeNameResult {
  result: ScopeNameResultNode
}

type ScopeNameResultNode {
  """
  Result Field title
  Result Field description
  """
  field: String!

  """
  Named Field title
  Named Field description
  """
  namedField: String!
}"
`;

exports[`schema.types generateUseCaseFieldConfig for profile fixture creates field config with arguments, resolver and description 1`] = `
{
  "args": {
    "input": {
      "description": "Use-case inputs",
      "type": "ScopeNameUseCaseInput",
    },
    "provider": {
      "type": "ScopeNameProviderOption",
    },
  },
  "description": "UseCase
UseCase description",
  "resolve": [Function],
  "type": "ScopeNameUseCaseResult",
}
`;

exports[`schema.types generateUseCaseFieldConfig for profile fixture with empty input fields creates arguments without input 1`] = `
{
  "args": {
    "provider": {
      "type": "ScopeNameProviderOption",
    },
  },
  "description": undefined,
  "resolve": [Function],
  "type": "ScopeNameUseCaseResult",
}
`;

exports[`schema.types generateUseCaseProviderParametersFields generates provider parameters for configured providers 1`] = `
{
  "accessToken": {
    "description": "Parameter accepted by superface",
    "type": "String",
  },
}
`;

exports[`schema.types generateUseCaseSecurityFields generates security fields for configured providers 1`] = `
{
  "api_key": {
    "description": "Security accepted by superface",
    "type": "SuperfaceApiKeySecurityValues",
  },
  "basic": {
    "description": "Security accepted by superface",
    "type": "SuperfaceBasicSecurityValues",
  },
  "bearer_token": {
    "description": "Security accepted by superface",
    "type": "SuperfaceBearerTokenSecurityValues",
  },
  "digest": {
    "description": "Security accepted by superface",
    "type": "SuperfaceDigestSecurityValues",
  },
}
`;

exports[`schema.types inputType returns GraphQLInputType 1`] = `
"type MyObjectType {
  """
  Primitive
  String
  """
  primitive_string: String

  """
  Primitive
  Number
  """
  primitive_number: Int

  """
  Primitive
  boolean
  """
  primitive_boolean: Boolean

  """
  Object
  Nested
  """
  nested_object: MyObjectTypeNestedObject
  list: [String!]

  """
  Enum title
  Enum description
  """
  enum: MyObjectTypeEnum
}

type MyObjectTypeNestedObject {
  """
  Nested field title
  Nested field description
  """
  field: String
}

"""
Enum title
Enum description
"""
enum MyObjectTypeEnum {
  ONE
  TWO
}"
`;

exports[`schema.types outputType returns GraphQLObjectType 1`] = `
"type MyObjectType {
  """
  Primitive
  String
  """
  primitive_string: String

  """
  Primitive
  Number
  """
  primitive_number: Int

  """
  Primitive
  boolean
  """
  primitive_boolean: Boolean

  """
  Object
  Nested
  """
  nested_object: MyObjectTypeNestedObject
  list: [String!]

  """
  Enum title
  Enum description
  """
  enum: MyObjectTypeEnum
}

type MyObjectTypeNestedObject {
  """
  Nested field title
  Nested field description
  """
  field: String
}

"""
Enum title
Enum description
"""
enum MyObjectTypeEnum {
  ONE
  TWO
}"
`;

exports[`schema.types scalarType returns GraphQLScalarType 1`] = `
""""
Scalar title
Scalar description
"""
scalar TestScalar"
`;
